modules/ipc/include/hephaestus/ipc/zenoh/action_server.h

```diff
@@ -0,0 +1,222 @@
+//================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+
+#pragma once
+
+#include <atomic>
+#include <functional>
+#include <future>
+#include <optional>
+
+#include <magic_enum.hpp>
+
+#include "hephaestus/concurrency/message_queue_consumer.h"
+#include "hephaestus/ipc/common.h"
+#include "hephaestus/ipc/publisher.h"
+#include "hephaestus/ipc/zenoh/internal/action_server_client_helper.h"
+#include "hephaestus/ipc/zenoh/service.h"
+#include "hephaestus/ipc/zenoh/session.h"
+#include "hephaestus/ipc/zenoh/types/action_server_types.h"
+#include "hephaestus/ipc/zenoh/types/action_server_types_protobuf.h"  // NOLINT(misc-include-cleaner)
+
+namespace heph::ipc::zenoh {
+
+/// If an action server is already serving a request it will not accept new ones.
+/// RequestT needs to be copyable and one copy will be made of it.
+/// ActionServer needs two callbacks:
+/// - RequestCallback: to check if the request is valid and if it can be served.
+/// - ExecuteCallback: to execute the request and send status updates.
+///   - Note that the execute callback is executed in a dedicated thread.
+/// Implementation details:
+/// - ActionServer contains a `Service` to receive the requests and a `MessageQueueConsumer` to execute them.
+/// - When a new request is accepted and started a new `Publisher` is created to send status updates.
+/// - The publisher is passed to `execute_cb`, which the user can use to publish status update of the
+///   execution.
+///   - The client when calling the server will create a temporary `Subscriber` to receive the updates.
+/// - When `execute_cb` finishes the final response is sent to the client via a `Service` created by the
+///   caller.
+/// NOTE: The reason why we can only process one request at a time is that with the current implementation the
+/// response `Service` and the update `Publisher` uses a topic name which just depends on the input topic
+/// name. This means that if we have multiple requests we will have multiple `Publisher` and `Subscriber` with
+/// the same topic name, which will cause conflicts. This could be solved by adding a unique identifier to the
+/// topic name, but this is not implemented yet.
+template <typename RequestT, typename StatusT, typename ReplyT>
+class ActionServer {
+public:
+  using RequestCallback = std::function<ActionServerRequestStatus(const RequestT&)>;
+  // NOTE: execution function is noexcept to avoid exceptions in the callback.
+  using ExecuteCallback = std::function<ReplyT(const RequestT&, Publisher<StatusT>&)>;
```

Comment:
Add also a stop request flag, so the user knows that they should stop the execution